### PR TITLE
Introduce a --syntaxOnly server mode

### PR DESCRIPTION
**Pull request by amcasey**
1. Disable the LS in all projects
 2. Don't create Program objects

Replaces #23037 & #23033

### CODE REVIEW

**Comment by amcasey** [comment]
I'm not thrilled about opening the config files because it creates some extra file watchers (for the config files themselves) and makes it harder for the editor to set flags, esp `noLib`, `noResolve`, `noEmit`, and `types`.

File: src/server/editorServices.ts [hunk 1]
```diff
@@ -1388,6 +1397,7 @@ namespace ts.server {
                 this.documentRegistry,
                 compilerOptions,
                 /*lastFileExceededProgramSize*/ this.getFilenameForExceededTotalSizeLimitForNonTsFiles(projectFileName, compilerOptions, files, externalFilePropertyReader),
+                /*languageServicePermanentlyDisabled*/this.syntaxOnly,
```
Comment (sheetalkamat) [inline src/server/editorServices.ts]: why do you need to pass this to each project, that should be checked by each project by just checking this.projectService.syntaxOnly?

Comment (amcasey) [reply src/server/editorServices.ts]: Just an oversight.

File: src/server/project.ts [hunk 2]
```diff
@@ -875,10 +878,12 @@ namespace ts.server {
             this.dirty = false;
             this.resolutionCache.finishCachingPerDirectoryResolution();
 
+            Debug.assert(oldProgram === undefined || this.program !== undefined);
+
             // bump up the version if
             // - oldProgram is not set - this is a first time updateGraph is called
             // - newProgram is different from the old program and structure of the old program was not reused.
-            const hasChanges = !oldProgram || (this.program !== oldProgram && !(oldProgram.structureIsReused & StructureIsReused.Completely));
+            const hasChanges = this.program && (!oldProgram || (this.program !== oldProgram && !(oldProgram.structureIsReused & StructureIsReused.Completely)));
```
Comment (sheetalkamat) [inline src/server/project.ts]: why would you be here when this.program === undefined ?

File: src/server/editorServices.ts [hunk 3]
```diff
@@ -1197,6 +1201,11 @@ namespace ts.server {
         private forEachConfigFileLocation(info: ScriptInfo,
             action: (configFileName: NormalizedPath, canonicalConfigFilePath: string) => boolean | void,
             projectRootPath?: NormalizedPath) {
+
+            if (this.syntaxOnly) {
+                return undefined;
```
Comment (sheetalkamat) [inline src/server/editorServices.ts]: I thought the idea was to create program but not bind it. Are we not going to do that. This means if file opened is from configured project, it wont pick its compiler options?

File: tests/cases/compiler/APISample_watcher.ts [hunk 4]
```diff
@@ -1,114 +0,0 @@
-﻿// @module: commonjs
```
Comment (mhegazy) [inline tests/cases/compiler/APISample_watcher.ts]: Do not think it is intentional to remove this file.

**Pull request merged**